### PR TITLE
add initial codes and figures

### PR DESCRIPTION
**Pull request by huqiwen0313**
Some initial codes for VAE and performance evaluation. All the codes are in the script folder.

### CODE REVIEW

File: single_cell_analysis/scripts/00-data.download.processing/.Rhistory [hunk 1]
```diff
@@ -0,0 +1,293 @@
+getwd()
```
Comment (gwaybio) [inline single_cell_analysis/scripts/00-data.download.processing/.Rhistory]: `.Rhistory` should be moved to the `.gitignore` file

File: single_cell_analysis/scripts/00-data.download.processing/data.download.processing.sh [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#
+# Usage: bash data.download.processing.sh
+#
+# Output: log(or TPM) -normalized count files and files contain celltype information. 
+#         files are located at data/real_data for real datasets and data/simulate_data for simulated datasets
+
+
+
+# 1. Generate simulated datasets
+  Rscript splatter.simulation.R
```
Comment (gwaybio) [inline single_cell_analysis/scripts/00-data.download.processing/data.download.processing.sh]: extra spaces in beginning

File: single_cell_analysis/scripts/00-data.download.processing/data.download.processing.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+
+# 2. download and process real datasets
+  Rscript real.datasets.download.R
```
Comment (gwaybio) [inline single_cell_analysis/scripts/00-data.download.processing/data.download.processing.sh]: here too

File: single_cell_analysis/scripts/00-data.download.processing/data.download.processing.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+
+# Pipeline to generate/download simulated and real single cell datasets and do data processing
+#
+# Usage: bash data.download.processing.sh
```
Comment (gwaybio) [inline single_cell_analysis/scripts/00-data.download.processing/data.download.processing.sh]: what directory should this file be run in?

File: single_cell_analysis/scripts/00-data.download.processing/splatter.simulation.R [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  # Args:
+  #  ncells: # of cells in each sample, default = 600
+  #  nGenes: # of genes/features in each sample, defalut = 10000
+  #  celltypes: # of celltypes, default = 5
+  #  bathsize: # of bathsize, default = 1
+  #  out.prob:  probability of a gene that become an outlier, default = 0.05
+  # Returns:
+  #  matrix of gene counts/expression and celltype information
+  
+  # caculate the probabilities that cells come from particular celltypes - currently even distributed
```
Comment (gwaybio) [inline single_cell_analysis/scripts/00-data.download.processing/splatter.simulation.R]: typo `calculate`

File: single_cell_analysis/scripts/00-data.download.processing/splatter.simulation.R [hunk 6]
```diff
@@ -0,0 +1,121 @@
+# Qiwen Hu 2018
```
Comment (gwaybio) [inline single_cell_analysis/scripts/00-data.download.processing/splatter.simulation.R:1]: Many lines in this file are quite long. Can you split long lines onto multiple lines?

File: single_cell_analysis/scripts/00-data.download.processing/splatter.simulation.R [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+seed = 628540
+set.seed(seed)
+
+simulation <- function(ncells = 600, nGenes = 10000, celltypes = 5, batchsize = 1, out.prob = 0.05, seed = 628540){
+  # Generate simulated single cell data from splatter
+  # Args:
+  #  ncells: # of cells in each sample, default = 600
+  #  nGenes: # of genes/features in each sample, defalut = 10000
+  #  celltypes: # of celltypes, default = 5
+  #  bathsize: # of bathsize, default = 1
```
Comment (gwaybio) [inline single_cell_analysis/scripts/00-data.download.processing/splatter.simulation.R]: typo

File: single_cell_analysis/scripts/celltype.vis.R [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+# visulization for real data
+input.dir <- file.path("features", "real_data")
+output.dir <- file.path("figures", "features_vis", "real_datasets")
+
+# list of real files
+real.file.list <- read.table(file.path(input.dir, "file.list.txt"))
+celltye_vis(input.dir, output.dir, real.file.list, option = "real")
+
+
```
Comment (gwaybio) [inline single_cell_analysis/scripts/celltype.vis.R]: Remove 3 extra lines

File: single_cell_analysis/scripts/pf.eval.kmeans.R [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+  
+  for(i in 1:length(tybalt_d1_files)){
+    filelist <- c(tybalt_d1_files[i], tybalt_d2_files[i], tybalt_d3_files[i], tsne_files[i],
+                  umap_files[i], zifa_files[i], pca_files[i])
+    # read files
+    datalist = lapply(file.path(input.dir, filelist), function(x) read.table(x, header=T)) 
+    cellinfo <- read.table(file.path(input.dir, cellinfo.files[i]), sep = "\t", header = TRUE)
+    
+    # extract dataset information
+    #dataset <- get_param(tybalt_d1_files[i], option = option)
```
Comment (gwaybio) [inline single_cell_analysis/scripts/pf.eval.kmeans.R]: remove?

File: single_cell_analysis/scripts/util/util.R [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    feature <- feature[, -1]
+  } 
+  
+  # convert celltype into numbers
+  celltype_num <- as.numeric(celltype)
+  sample_id <- seq(1:nrow(feature))
+  
+  # iterative k-means
+  # set number of iteration to be 50
+  iter <- 50
```
Comment (gwaybio) [inline single_cell_analysis/scripts/util/util.R]: worth making this a function argument?

File: single_cell_analysis/scripts/util/util.R [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+  
+  return(data.frame(nmi = round(mean(nmi_score_all), 2), ari = round(mean(ari_score_all), 2), 
+                    nmi_var = round(sd(nmi_score_all), 2), ari_var = round(sd(ari_score_all), 2)))
+}
+
+dim_reduc <- function(data, option = "SIMLR", seed = 12345){
+  # Perform SIMLR or PCA dimension reduction 
+  # Args:
+  #   data: matrix used to do dimension reduction
+  #   option: SILMR/PCA
```
Comment (gwaybio) [inline single_cell_analysis/scripts/util/util.R]: typo? Can you also add a bit of documentation of what SIMLR is?

File: single_cell_analysis/scripts/util/util.R [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+  #  Rank-based criteria. Neurocomputing 72
+  #
+  # Args:
+  #   count.matrix: orignal count.matrix (normalized) before dimension reduction
+  #   dm.matrix: low dimension matrix after dimension reduction
+  # Returns:
+  #   q.local and q.global
+  
+  coranking.matrix <- coRanking::coranking(t(count.matrix), dm.matrix[, 2:3])
+  numeric(nrow(coranking.matrix))
```
Comment (gwaybio) [inline single_cell_analysis/scripts/util/util.R]: what is this line doing?

File: single_cell_analysis/scripts/util/util.R [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+  
+  # average silhouette value
+  ss.mean <- mean(ss[, 3])
+  return(round(ss.mean, 2))
+}
+
+coranking_metric <- function(count.matrix, dm.matrix){
+  # Compute coranking metric to evaluate the performance of dimension reduction
+  # for detail definition of coranking matrix, see 
+  # Lee, J.A., Lee, J.A., Verleysen, M., 2009. Quality assessment of dimensionality reduction:
```
Comment (gwaybio) [inline single_cell_analysis/scripts/util/util.R]: This is interesting. Can you add a bit of intuition of how to interpret this coranking metric?

File: single_cell_analysis/scripts/util/vis.R [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+    for(i in 1:length(par)){
+      # get performance metric for a specific parameter
+      param_perf <- cluster_perf[grep(par[i], cluster_perf$files), ]
+      param_perf$files <- gsub(paste(par[i], ".", sep = ""), 
+                             "", param_perf$files)
+      
+      param_perf <- param_perf[order(as.numeric(param_perf$files)), ]
+      
+      pdf(file.path(output.dir, paste(par[i], type, "pdf", sep = ".")), width=7, height=5)
+      my_palette <- colorRampPalette(c("white", "blue"))(n = 50)
```
Comment (gwaybio) [inline single_cell_analysis/scripts/util/vis.R]: don't need to define this palette through ever loop

File: single_cell_analysis/scripts/util/vis.R [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+  # add celltype information into file
+  if(data == "simulated"){
+    file <- cbind(file, celltype$Group)
+  } else{
+    file <- cbind(file, celltype$celltype)
+  }
+  colnames(file) <- c("id", "X1", "X2", "Group")
+    
+  # remove outlier for visulization
+  file <- file[file$X1 > -50 & file$X2 > -50, ]
```
Comment (gwaybio) [inline single_cell_analysis/scripts/util/vis.R]: how are these parameters defined?

**Review by gwaybio — COMMENTED** [review]
A couple comments. looks mostly good, just need some clarification

File: single_cell_analysis/scripts/00-data.download.processing/data.download.processing.sh (hunk 4 above)
Comment (huqiwen0313) [reply single_cell_analysis/scripts/00-data.download.processing/data.download.processing.sh]: scripts/00-data.download.processing.   Added path to the usage line

File: single_cell_analysis/scripts/pf.eval.kmeans.R (hunk 9 above)
Comment (huqiwen0313) [reply single_cell_analysis/scripts/pf.eval.kmeans.R]: yea, I am not suing this function. I deleted this line.

File: single_cell_analysis/scripts/util/util.R (hunk 12 above)
Comment (huqiwen0313) [reply single_cell_analysis/scripts/util/util.R]: This should be lcmc <- numeric(...)... define lcmc as a numeric vector. Thanks for catching it.

File: single_cell_analysis/scripts/util/vis.R (hunk 15 above)
Comment (huqiwen0313) [reply single_cell_analysis/scripts/util/vis.R]: This is only for visualization purpose to make the figure looks better. The parameters are defined based on the distribution of X and Y values. There is only 1 -2 points are located outside of this range. I added more description to clarify.

File: single_cell_analysis/scripts/celltype.vis.R [hunk 16]
```diff
@@ -57,15 +57,10 @@ output.dir <- file.path("figures", "features_vis", "simulated_data")
...
 output.dir <- file.path("figures", "features_vis", "real_datasets")
 
 # list of real files
 real.file.list <- read.table(file.path(input.dir, "file.list.txt"))
-celltye_vis(input.dir, output.dir, real.file.list, option = "real")
-
-
-
-
+celltye_vis(input.dir, output.dir, real.file.list, option = "real")
```
Comment (gwaybio) [inline single_cell_analysis/scripts/celltype.vis.R]: the red arrow means you need to add a new line. There should be 1 empty line at the end of file

**Comment by huqiwen0313** [comment]
@gwaygenomics, I addressed all the comments, can you please take a look at it again ? Thanks

File: single_cell_analysis/scripts/00-data.download.processing/data.download.processing.sh [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+# Pipeline to generate/download simulated and real single cell datasets and do data processing
+#
+# Usage: bash scripts/00-data.download.processing/data.download.processing.sh
+#
+# Output: log(or TPM) - normalized count files and files contain celltype information. 
+#         files are located at data/real_data for real datasets and data/simulate_data for simulated datasets
+
+# 1. Generate simulated datasets
+  Rscript splatter.simulation.R
```
Comment (gwaybio) [inline single_cell_analysis/scripts/00-data.download.processing/data.download.processing.sh]: If this is supposed to be run in `scripts/00-data.download.processing/` I do not think these lines will work. Should the user first navigate into this folder and then execute the bash script? Alternatively can add the full path to the R script (and all other scripts below)

**Review by gwaybio — APPROVED** [review]
No need for me to re-review

File: single_cell_analysis/scripts/00-data.download.processing/data.download.processing.sh (hunk 17 above)
Comment (huqiwen0313) [reply single_cell_analysis/scripts/00-data.download.processing/data.download.processing.sh]: Thanks ! added full path to the R script

**Pull request merged**